### PR TITLE
Fix scoreboard sorting failing on custom scoreboardOrderSubmissions, add more error checking when creating scoreboard entries / displaying

### LINKED ISSUE #2086: ArgumentError when viewing scoreboard
**Your environment**
 Carnegie Mellon's Autolab

**Steps To Reproduce**
Reported but unable to reproduce.

**Stacktrace**
<details>

``````
An ArgumentError occurred in scoreboards#show:

comparison of Hash with Hash failed
app/controllers/scoreboards_controller.rb:107:in `sort'

-------------------------------
Request:
-------------------------------

* URL        : https://autolab.andrew.cmu.edu/courses/15213-s24/assessments/attacklab/scoreboard
* HTTP Method: GET
* Parameters : {"controller"=>"scoreboards", "action"=>"show", "course_name"=>"15213-s24", "assessment_name"=>"attacklab"}
* Timestamp  : 2024-02-09 23:04:30 -0500
* Server : [autolab-11.andrew.cmu.edu](http://autolab-11.andrew.cmu.edu/)
* Rails root : /opt/autolab
* Process: 2834767

-------------------------------
Environment:
-------------------------------

* AUTOCONFIG_COURSE_DIR                                   : /opt/autolab/courses
* DEVISE_SECRET_KEY                                       : [FILTERED]
* GITHUB_CLIENT_ID                                        : 5f52e84b5516f04cf97e
* GITHUB_CLIENT_SECRET                                    : [FILTERED]
* HTTPS                                                   : on
* HTTP_ACCEPT                                             : text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7
* HTTP_ACCEPT_ENCODING                                    : gzip, deflate, br
* HTTP_ACCEPT_LANGUAGE                                    : en-US,en;q=0.9
* HTTP_CACHE_CONTROL                                      : max-age=0
* HTTP_COOKIE                                             : [FILTERED]
* HTTP_HOST                                               : [autolab.andrew.cmu.edu](http://autolab.andrew.cmu.edu/)
* HTTP_IF_NONE_MATCH                                      : W/"04c678d9333c62557c0b5f5c92bf4ea5"
* HTTP_REFERER                                            : https://autolab.andrew.cmu.edu/courses/15213-s24/assessments/attacklab
* HTTP_SEC_CH_UA                                          : "Not_A Brand";v="8", "Chromium";v="120", "Google Chrome";v="120"
* HTTP_SEC_CH_UA_MOBILE                                   : ?0
* HTTP_SEC_CH_UA_PLATFORM                                 : "Windows"
* HTTP_SEC_FETCH_DEST                                     : document
* HTTP_SEC_FETCH_MODE                                     : navigate
* HTTP_SEC_FETCH_SITE                                     : same-origin
* HTTP_SEC_FETCH_USER                                     : ?1
* HTTP_UPGRADE_INSECURE_REQUESTS                          : 1
* HTTP_USER_AGENT                                         : Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/[120.0.0.0](http://120.0.0.0/) Safari/537.36
* HTTP_VERSION                                            : HTTP/1.1
* LOCKBOX_MASTER_KEY                                      : [FILTERED]
* LTI_TOOL_PRIVATE_KEY                                    : [FILTERED]
* ORIGINAL_FULLPATH                                       : /courses/15213-s24/assessments/attacklab/scoreboard
* ORIGINAL_SCRIPT_NAME                                    :
* PASSENGER_CONNECT_PASSWORD                              : [FILTERED]
* PATH_INFO                                               : /courses/15213-s24/assessments/attacklab/scoreboard
* QUERY_STRING                                            :
* REQUEST_METHOD                                          : GET
* REQUEST_URI                                             : /courses/15213-s24/assessments/attacklab/scoreboard
* RESTFUL_HOST                                            : [autolab-tango.andrew.cmu.edu](http://autolab-tango.andrew.cmu.edu/)
* RESTFUL_KEY                                             : [FILTERED]
* RESTFUL_PORT                                            : 443
* ROUTES_8020_SCRIPT_NAME                                 :
* SCRIPT_NAME                                             :
* SECRET_KEY_BASE                                         : [FILTERED]
* SERVER_NAME                                             : [autolab.andrew.cmu.edu](http://autolab.andrew.cmu.edu/)
* SERVER_PORT                                             : 443
* SERVER_PROTOCOL                                         : HTTP/1.1
* SERVER_SOFTWARE                                         : Apache/2.4.37 (Red Hat Enterprise Linux) OpenSSL/1.1.1k Phusion_Passenger/6.0.19
* action_controller.instance                              : #<ScoreboardsController:0x00007f6fbe2660a8>
* action_dispatch.authenticated_encrypted_cookie_salt     : [FILTERED]
* action_dispatch.backtrace_cleaner                       : #<Rails::BacktraceCleaner:0x00007f6fbff1ece0>
* action_dispatch.content_security_policy                 :
* action_dispatch.content_security_policy_nonce_directives:
* action_dispatch.content_security_policy_nonce_generator :
* action_dispatch.content_security_policy_report_only     : false
* action_dispatch.cookies                                 : [FILTERED]
* action_dispatch.cookies_digest                          : [FILTERED]
* action_dispatch.cookies_rotations                       : [FILTERED]
* action_dispatch.cookies_same_site_protection            : [FILTERED]
* action_dispatch.cookies_serializer                      : [FILTERED]
* action_dispatch.encrypted_cookie_cipher                 : [FILTERED]
* action_dispatch.encrypted_cookie_salt                   : [FILTERED]
* action_dispatch.encrypted_signed_cookie_salt            : [FILTERED]
* action_dispatch.http_auth_salt                          : [FILTERED]
* action_dispatch.key_generator                           : #<ActiveSupport::CachingKeyGenerator:0x00007f6fc6f0d1f0>
* action_dispatch.logger                                  : #<ActiveSupport::Logger:0x00007f6fbfff3918>
* action_dispatch.parameter_filter                        : [:password, :password, :session, :warden, :secret, :salt, :cookie, :csrf, :restful_key, :lockbox_master_key, :lti_tool_private_key, /^((?-mix:client_secret|authentication_token|access_token|refresh_token|code))$/]
* action_dispatch.permissions_policy                      :
* action_dispatch.redirect_filter                         : []
* action_dispatch.request.content_type                    :
* action_dispatch.request.formats                         : [#<Mime::Type:0x00007f6fc91cf620 @synonyms=["application/xhtml+xml"], @symbol=:html, @string="text/html", @hash=3419628624715560641>]
* action_dispatch.request.parameters                      : {"controller"=>"scoreboards", "action"=>"show", "course_name"=>"15213-s24", "assessment_name"=>"attacklab"}
* action_dispatch.request.path_parameters                 : {:controller=>"scoreboards", :action=>"show", :course_name=>"15213-s24", :assessment_name=>"attacklab"}
* action_dispatch.request.query_parameters                : {}
* action_dispatch.request.request_parameters              : {}
* action_dispatch.request.unsigned_session_cookie         : [FILTERED]
* action_dispatch.request_id                              : ea20d248-e1fa-4734-8996-266bc02fc004
* action_dispatch.routes                                  : #<ActionDispatch::Routing::RouteSet:0x00007f6fc00084a8>
* action_dispatch.secret_key_base                         : [FILTERED]
* action_dispatch.show_detailed_exceptions                : false
* action_dispatch.show_exceptions                         : true
* action_dispatch.signed_cookie_digest                    : [FILTERED]
* action_dispatch.signed_cookie_salt                      : [FILTERED]
* action_dispatch.use_authenticated_cookie_encryption     : [FILTERED]
* action_dispatch.use_cookies_with_metadata               : [FILTERED]
* https_proxy                                             : http://proxy.andrew.cmu.edu:3128/
* no_proxy                                                : [cmu.edu](http://cmu.edu/),cmu.local
* rack.attack.called                                      : true
* rack.attack.match_type                                  : safelist
* rack.attack.matched                                     : allow from localhost
* rack.errors                                             : #<IO:0x00007f6fcf60fb80>
* rack.hijack                                             : #<Proc:0x00007f6fbe289828 /opt/autolab/vendor/bundle/ruby/3.1.0/gems/passenger-6.0.19/src/ruby_supportlib/phusion_passenger/rack/thread_handler_extension.rb:94 (lambda)>
* rack.hijack?                                            : true
* rack.input                                              : #<PhusionPassenger::Utils::TeeInput:0x00007f6fbe2898c8>
* rack.multiprocess                                       : true
* rack.multithread                                        : false
* rack.request.cookie_hash                                : [FILTERED]
* rack.request.cookie_string                              : [FILTERED]
* rack.request.query_hash                                 : {}
* rack.request.query_string                               :
* rack.run_once                                           : false
* rack.session                                            : [FILTERED]
* rack.session.options                                    : [FILTERED]
* rack.tempfiles                                          : []
* rack.url_scheme                                         : https
* rack.version                                            : [1, 3]
* warden                                                  : [FILTERED]

-------------------------------
Backtrace:
-------------------------------

app/controllers/scoreboards_controller.rb:107:in `sort'
app/controllers/scoreboards_controller.rb:107:in `show'

-------------------------------
Data:
-------------------------------

* data: {:user=>
#<User id: 96965, [REDACTED]>,
:course=>
#<Course:0x00007f6fbe26e528
id: 2162,
name: "15213-s24",
semester: "s24",
late_slack: 300,
grace_days: 5,
display_name: "15213-s24: Introduction to Computer Systems",
start_date: Fri, 12 Jan 2024,
end_date: Thu, 16 May 2024,
disabled: false,
exam_in_progress: false,
version_threshold: -1,
late_penalty_id: 106154,
version_penalty_id: 106155,
cgdub_dependencies_updated_at:
Thu, 08 Feb 2024 18:17:12.000000000 EST -05:00,
gb_message: "",
website: "">,
:assessment=>
#<Assessment:0x00007f6fbe27dac8
id: 29104,
due_at: Thu, 15 Feb 2024 23:59:00.000000000 EST -05:00,
end_at: Sun, 18 Feb 2024 23:59:00.000000000 EST -05:00,
start_at: Thu, 08 Feb 2024 18:30:00.000000000 EST -05:00,
name: "attacklab",
description:
"\"I came out to have a good time and I'm honestly feeling so attacked right now\"",
created_at: Tue, 30 Jan 2024 22:41:07.000000000 EST -05:00,
updated_at: Thu, 08 Feb 2024 17:29:06.000000000 EST -05:00,
course_id: 2162,
display_name: "Attack Lab",
handin_filename: "handin.tar",
handin_directory: "handin",
max_grace_days: 1,
handout: "",
writeup: "writeup/attacklab.pdf",
allow_unofficial: false,
max_submissions: -1,
disable_handins: false,
exam: false,
max_size: 2,
version_threshold: -1,
late_penalty_id: nil,
version_penalty_id: nil,
has_svn: nil,
quiz: false,
quizData: "",
remote_handin_path: "",
category_name: "Lab",
group_size: 1,
embedded_quiz_form_data: nil,
embedded_quiz: false,
github_submission_enabled: false,
allow_student_assign_group: true,
is_positive_grading: false>,
:submission=>nil}
``````
</details>

### PR DESCRIPTION
**Pull request by 20wildmanj**
## Description

- Adds more error-catching logic when calling custom `scoreboardOrderSubmissions`, in the case that an exception is raised, and tells the instructor via an error message
- Adds a rescue block to the top-level sort call in `#show` so that if a bad sort / comparison occurs, the error can be caught
- Instead of returning scoreboard entries with "-" on errors, raise an error, and in `#show`, revert to an empty array
- In the `#show` view code, check that the current entry length is greater than the index in order to show a score, else show an error icon (which handles the empty array case)
- Fix up logging, add more specific errors to be returned to instructors, and redirect students back to assessments index

## Motivation and Context

Closes #2086 

## How Has This Been Tested?

Import the provided assessment:
[randomlab5_20240217.tar.zip](https://github.com/autolab/Autolab/files/14320229/randomlab5_20240217.tar.zip)

1. Upload at least two submissions as students
2. View scoreboard, see that it renders fine (if it is not rendering fine, upload `autograder.tar`as the autograder)
3. Upload the assessment config file `randomlab5-bad-scoreboard-order.rb` (in Edit Assessment you can upload)
  - this test returns a bad sort, causing an ArgumentError
5. See that you get an error as follows:
Screenshot 2024-02-17 at 3 22 59 PM
6. Try to view scoreboard as a student and see that you get redirected back to assessment index with an error flash
7. Upload the initial assessment config file back, and then upload `autograde-bad.tar` as the new autograder
 - this autograder returns just a number instead of an array
9. Regrade all submissions
10. view scoreboard as instructor, see that following error appears:
Screenshot 2024-02-17 at 3 59 06 PM
11. View scoreboard as student, see that scoreboard is visible but error icons appear:
Screenshot 2024-02-17 at 4 03 59 PM
12. Upload the old autograder.tar back, and regrade all
13. Then add this code after line 296 in `scoreboards_controller.rb`:
```
if [true, false].sample
      parsed["scoreboard"] = 0
    end
```
( you may need to also reload the config file)
  - this replicates the scenario where sometimes the scoreboard results aren't correct / autograder failed
14. view the scoreboard as an instructor, see the following error
Screenshot 2024-02-17 at 4 41 53 PM

15. As a student, see something like this (all the bad scoreboard entries are at the bottom)
Screenshot 2024-02-17 at 2 30 28 PM

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

File: app/views/scoreboards/show.html.erb [hunk 1]
```diff
@@ -61,7 +69,7 @@
               <% if c["img"] %>
                 <td><img src="data:image/png;base64,<%= grade[:entry][i] %>"></td>
               <% else %>
-                <% if grade[:entry].is_a?(Array) && grade[:entry][i] != "-" %>
+                <% if grade[:entry].is_a?(Array)  && grade[:entry].length > i %>
                   <td><%= grade[:entry][i] %></td>
                 <% else %>
                   <%= render 'error_icon' %>
```
[the remark this answers is not in the document]
Comment (damianhxy) [reply app/views/scoreboards/show.html.erb:75]: @20wildmanj This was introduced by #1982 and unfortunately is in production rn

Screenshot 2024-02-18 at 02 13 34

[the remark this answers is not in the document]
Comment (cg2v) [reply app/views/scoreboards/show.html.erb:75]: I removed the 'q' from the production instance

[the remark this answers is not in the document]
Comment (damianhxy) [reply app/views/scoreboards/show.html.erb:75]: Thanks! The symbol tree also doesn't seem to be working in prod, are the new systems missing ctags?

[the remark this answers is not in the document]
Comment (20wildmanj) [reply app/views/scoreboards/show.html.erb:75]: Added a commit to remove the typo, sorry for not catching this earlier!

File: app/controllers/scoreboards_controller.rb [hunk 2]
```diff
@@ -105,22 +105,68 @@ def show
...
       end
-    rescue StandardError => e
+    rescue ArgumentError => e
       if @cud.instructor?
         # not using flash because could be too large of a message to pass
-        @errorMessage = "An error occurred while calling "\
-          "scoreboardOrderSubmissions(#{a.inspect},\n"\
-          "#{b.inspect})"
+        @errorMessage = "An error occurred while sorting "\
+          "submissions . Please make sure your scoreboard results are returned " \
```
Comment (damianhxy) [inline app/controllers/scoreboards_controller.rb:159]: Nit: extra space before period

**Comment by damianhxy** [comment]
When updating the assessment rb file, if you go straight to the scoreboard (i.e. don't load the assessment index page), there is an error about an undefined method.

Screenshot 2024-02-21 at 02 19 53

However, after loading the assessment index page at least once, the error message is now as expected

Screenshot 2024-02-21 at 02 31 54


Suspect it is because `reload_config_file` is not called by `assessments_controller.rb#update`

File: app/controllers/scoreboards_controller.rb [hunk 3]
```diff
@@ -246,44 +289,33 @@ def createScoreboardEntry(scores, autoresult)
...
-      if !parsed["scoreboard"].is_a?(Array) && @cud.instructor?
-        flash[:error] = "Error parsing scoreboard for autograded assessment: scoreboard result is"\
-          " not an array. Please ensure that the autograder returns scoreboard results as an array."
-      end
+
+    parsed = ActiveSupport::JSON.decode(autoresult)
+
+    # ensure that the parsed result is a hash with scoreboard field, where scoreboard is an array
+    if (!parsed || !parsed.is_a?(Hash) || !parsed["scoreboard"] ||
+        !parsed["scoreboard"].is_a?(Array)) &&
```
Comment (damianhxy) [inline app/controllers/scoreboards_controller.rb:297]: Extraneous `&&`?

File: app/controllers/scoreboards_controller.rb [hunk 4]
```diff
@@ -246,44 +289,33 @@ def createScoreboardEntry(scores, autoresult)
...
-        # Give up and bail
-        return ["-"]
-      end
     end
+
+    # If there is no autoresult for this student (typically
+    # because their code did not compile or it segfaulted and
+    # the instructor's autograder did not catch it) then
+    # raise an error, will be handled by caller
+    raise StandardError if !parsed || !parsed.is_a?(Hash) || !parsed["scoreboard"] ||
```
Comment (damianhxy) [inline app/controllers/scoreboards_controller.rb:312]: Why not move this raise into the `if`-block above?

**Review by damianhxy — CHANGES_REQUESTED** [review]
Functionality generally looks good, but left some nits

File: app/views/scoreboards/show.html.erb [hunk 5]
```diff
@@ -5,6 +5,14 @@
   <%= javascript_include_tag "sorttable" %>
 <% end %>
 
+<% unless @errorMessage.nil? %>
+  <h4 class="error-header">Error Rendering Scoreboard:</h4>
+  <div>
+    <pre style="text-wrap: wrap;"><%= @errorMessage %></pre>
+    <pre><%= @error %></pre>
```
[the remark this answers is not in the document]
Comment (20wildmanj) [reply app/views/scoreboards/show.html.erb:12]: Removed the @error field

File: app/views/scoreboards/show.html.erb [hunk 6]
```diff
@@ -5,6 +5,14 @@
   <%= javascript_include_tag "sorttable" %>
 <% end %>
 
+<% unless @errorMessage.nil? %>
+  <h4 class="error-header">Error Rendering Scoreboard:</h4>
+  <div>
+    <pre style="text-wrap: wrap;"><%= @errorMessage %></pre>
```
[the remark this answers is not in the document]
Comment (20wildmanj) [reply app/views/scoreboards/show.html.erb:11]: Changed how the wrapping works:

Chrome:
Screenshot 2024-02-22 at 10 51 32 PM

Firefox:
Screenshot 2024-02-22 at 10 51 50 PM

Safari:
Screenshot 2024-02-22 at 10 52 32 PM

**Comment by 20wildmanj** [comment]
> When updating the assessment rb file, if you go straight to the scoreboard (i.e. don't load the assessment index page), there is an error about an undefined method.
> 
> Screenshot 2024-02-21 at 02 19 53
> 
> However, after loading the assessment index page at least once, the error message is now as expected
> 
> Screenshot 2024-02-21 at 02 31 54
> 
> Suspect it is because `reload_config_file` is not called by `assessments_controller.rb#update`

I noticed this while testing, not sure why going to index would resolve this though because doesn't seem that index doesn't do any calls to reload the config file? Not sure if it's worth looking into a fix for a hook that very few classes use (and the ones that do should understand how to reload the config file)

**Review by damianhxy — APPROVED** [review]
Ran through testing workflow, LGTM

**Pull request merged**